HeinOnline.js

```diff
@@ -0,0 +1,240 @@
+    (at your option) any later version.
+
+    Zotero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+    ***** END LICENSE BLOCK *****
+*/
+
+
+/*
+    ***************
+    ** Utilities **
+    ***************
+*/
+
+// Get any search results from current page
+// Used in detectWeb() and doWeb()
+function getSearchResults(doc, url) {
+	var results = doc.getElementsByClassName("lucene_search_result_b"),
+		items = {},
+		found = false
+	for (var i=0, ilen=results.length; i<ilen; i++) {
+		var url = getXPathStr("href", results[i], './/a[1]');
+
+		var title = getXPathStr("textContent", results[i], './/a[1]');
+		title = ZU.trimInternal(title);
+		//title = title.replace(/\s*\[[^\]]*\]$/, '');
+
+		if (!title || !url) continue;
+		
+		items[url] = title;
+		found = true;
+	}
+	return found ? items : false;
+}
+
+// Get the string value of the first object matching XPath
+function getXPathStr(attr, elem, path) {
+	var res = ZU.xpath(elem, path);
+	res = res.length ? res[0][attr] : '';
+	return res ? res : '';
+}
+
+// Extract query values to keys on an object
+function extractQueryValues(url) {
+	var ret = {};
+	ret.base = url.replace(/(.*?)[a-zA-Z]+\?.*/, "$1");
+	var query = url.replace(/.*?\?/, "");
+	query = query.split("&");
+	for (var i=0,ilen=query.length;i<ilen;i++) {
+		let pair = query[i].split("=");
+		ret[pair[0]] = pair[1];
+	}
+	return ret;
+}
+
+// Remangle HeinOnline RIS to form expected by translator
+function fixRIS (ris) {
+	var lines = ris.split("\n");
+	for (var i=0,ilen=lines.length;i<ilen;i++) {
+		lines[i] = lines[i].replace(/^([A-Z0-9][A-Z0-9]) /, "$1  - ");
+		lines[i] = lines[i].replace(/^VO  /, "VL  ");
+		lines[i] = lines[i].replace(/^YR  /, "DA  ");
+		lines[i] = lines[i].replace(/^OP  /, "EP  ");
+	}
+	ris = lines.join("\n");
+	return ris;
+}
+
+// Not all pages have a downloadable PDF
+function translateRIS(ris, pdfURL) {
+	var trans = Zotero.loadTranslator('import');
+	trans.setTranslator('32d59d2d-b65a-4da4-b0a3-bdd3cfb979e7');//https://github.com/zotero/translators/blob/master/RIS.js
+	trans.setString(ris);
+	if (pdfURL) {
+		trans.setHandler('itemDone', function (obj, item) {
+			item.attachments = [{
+				title: "Full Text PDF",
+				url: pdfURL,
+				mimeType: "application/pdf"
+			}];
+			item.complete();
+		});
+	} else {
+		trans.setHandler('itemDone', function (obj, item) {
+			item.complete();
+		});
+	}
+	trans.translate();
+}
+
+function translateCOinS(COinS) {
+    var item = new Zotero.Item();
+    try {
+        Zotero.Utilities.parseContextObject(COinS, item);
+        item.complete();
+    } catch(e) {
+        Zotero.debug("error parsing COinS");
+    }
```

Comment:
Let this throw, so the user knows that there was an error and so that we get a report